### PR TITLE
Secret rotation watcher

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR introduces a watcher that monitors changes in secret rotations. When a user establishes a rotation policy for a secret, the SQL trigger we've implemented activates the change stream infrastructure, which in turn triggers an event in the secret rotation watcher within the secret service. This watcher retrieves the secret's URI, its current revision, and the next scheduled rotation time, and then forwards this information to the consumer side.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```
juju exec --unit dummy-source/0 "secret-add data=foo"
secret://07fa70f6-9622-4fee-8f9b-50df07e6ddd9/cp85bfiit0018vbp0ej0

juju exec --unit dummy-source/0 "secret-set cp85bfiit0018vbp0ej0 --rotate hourly"

juju show-status-log dummy-source/0
Time                        Type       Status       Message
...
24 May 2024 18:54:23+10:00  workload   maintenance  Started
24 May 2024 18:54:23+10:00  juju-unit  idle
24 May 2024 18:54:45+10:00  juju-unit  executing    running action juju-exec
24 May 2024 18:54:45+10:00  juju-unit  idle
24 May 2024 18:54:55+10:00  juju-unit  executing    running action juju-exec
24 May 2024 18:54:55+10:00  juju-unit  idle
24 May 2024 18:55:25+10:00  juju-unit  executing    running secret-rotate hook for secret:cp85bfiit0018vbp0ej0
24 May 2024 18:55:25+10:00  juju-unit  idle


juju debug-log --color --replay -m model-secrets-offer
unit-dummy-source-0: 18:55:25 INFO juju.worker.uniter.operation skipped "secret-rotate" hook (missing)
```

## Documentation changes

No

## Links

**Jira card:** JUJU-5894

### CODE REVIEW

**Comment by ycliuhw** [comment]
> Can we get a description on this please?

Sorry, this one is not ready to review. I should have made it as a drap PR.
It seems the code owner stuff automatically requested the reviewers.

File: domain/secret/state/state_test.go [hunk 1]
```diff
@@ -3191,6 +3191,138 @@ func (s *stateSuite) TestGetRemoteConsumedSecretURIsWithChangesFromOfferingSide(
...
+		},
+		{
+			URI:             uri2,
+			Revision:        2,
+			NextTriggerTime: now.Add(2 * time.Hour).UTC(),
+		},
+	})
+
+	result, err = st.GetSecretsRotationChanges(ctx,
+		domainsecret.ApplicationOwners{"mysql", "mediawiki"}, domainsecret.UnitOwners{"mysql/0", "mediawiki/0"},
```
Comment (wallyworld) [inline domain/secret/state/state_test.go:3276]: The test set up should ensure there are other secrets owned by these units that are not uri1

File: domain/secret/state/state_test.go [hunk 2]
```diff
@@ -3191,6 +3191,138 @@ func (s *stateSuite) TestGetRemoteConsumedSecretURIsWithChangesFromOfferingSide(
...
+	c.Check(err, jc.ErrorIsNil)
+	c.Check(result, jc.SameContents, []domainsecret.RotationInfo{
+		{
+			URI:             uri2,
+			Revision:        2,
+			NextTriggerTime: now.Add(2 * time.Hour).UTC(),
+		},
+	})
+
+	result, err = st.GetSecretsRotationChanges(ctx, domainsecret.ApplicationOwners{"mysql"}, nil)
```
Comment (wallyworld) [inline domain/secret/state/state_test.go:3301]: Same here for mysql app

**Review by wallyworld — APPROVED** [review]
A few things I think can be tweaked. I think the test set up should add more data to test that it is properly excluded when running the queries.
The comments here also probably apply to the existing obsolete revision code.

File: domain/secret/service/service.go [hunk 3]
```diff
@@ -420,13 +430,13 @@ func (s *SecretService) ListCharmSecrets(ctx context.Context, owners ...CharmSec
 }
 
 // GetSecret returns the secret with the specified URI.
-// If returns [secreterrors.SecretNotFound] is there's no such secret.
+// If returns [secreterrors.SecretNotFound] is there's no such domainsecret.
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply domain/secret/service/service.go:433]: hm, this was a repalceAll for imports. Fixed!

File: domain/secret/service/service_test.go [hunk 4]
```diff
@@ -1609,3 +1609,86 @@ func (s *serviceSuite) TestWatchRemoteConsumedSecretsChanges(c *gc.C) {
...
+		},
+		CharmSecretOwner{
+			Kind: UnitOwner,
+			ID:   "mysql/1",
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(w, gc.NotNil)
+	defer workertest.CleanKill(c, w)
+	wC := watchertest.NewSecretsTriggerWatcherC(c, w)
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply domain/secret/service/service_test.go:1673]: The above defer ensures that the raw watcher is closed peacefully.
We already do the AssertNoChange in the wc watcher.

**Pull request merged**